text/0000-sigstore-integration.md

```diff
@@ -0,0 +1,387 @@
+
+The crates.io sub-flow on publish is described below:
+
+![crates.io flow](https://raw.githubusercontent.com/lulf/rfc-resources/main/sigstore/cargo_sigstore-crates.io.drawio.png)
+
+The following components need changing for crates.io:
+
+1. Additional columns must be added to the PostgreSQL database
+1. HTTP API must handle a new version of NewCratePublish request with the signature data
+1. The HTTP handler talks to Rekor and verifies that the signature is valid and that owner matches certificate
+
+## Cargo verify flow
+
+The cargo flow during verification is shown below:
+
+![verify flow](https://raw.githubusercontent.com/lulf/rfc-resources/main/sigstore/cargo_sigstore-verify.drawio.png)
+
+The following changes must be made to cargo when building/verifying a crate:
+
+1. Retrieve the crate contents, signature and certificate.
+1. Verify that the crate contents is signed by the signature.
+1. Verify that the certificate and signature is present in the transparency log.
+
+### Offline flow
+
+The offline verification flow is shown below
+
+![verify flow](https://raw.githubusercontent.com/lulf/rfc-resources/main/sigstore/cargo_sigstore-verify-offline.drawio.png)
+
+## Note on cargo dependencies
+
+The interaction with Sigstore services is implemented in the `sigstore-rs` crate, which has the following sub-dependencies:
+
+* tough: used to fetch fulcio public key and rekor certificates from Sigstore's TUF repository. 
+* openid-connect: this is used to generate keyless signatures via Rust
+* oauth2: this is a transitive dependency of openid-connect
+
+These crate use `reqwest` and some async Rust. To handle this, we can either:
+
+* Add the necessary dependencies to cargo with the goal of replacing the curl with reqwest to avoid multiple HTTP clients in use
+* Use the lower level APIs in sigstore-rs and implement the HTTP interaction using curl which is used by cargo today. 
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Cargo may be slower if signing and verification is enabled by default. To start, it should be made opt-in for a period so that maintainers and users can get to know the tool and we can learn more about the reliability.
+
+Reliance on the public [Sigstore uptime](https://www.chainguard.dev/unchained/sigstore-is-generally-available) of 99.5% availability goal may affect users signing and verifying crates, as well as crates.io verifying signature identities upon publish.
+
+Additional metadata will be stored for each crate, which will increase storage requirements for the index.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+There are other approaches to key management out there, such as GPG. However, these typically have a higher effort requiring keys to be managed without the downsides of longevity and compromises. Sigstore is meant to capture some of these common patterns of managing keys and storing information for auditing, which lowers the barrier for signing crates.
```

Comment:
The downsides of... compromises? That seems to just say that there are downsides, not what they are.

Additionally, longevity can be seen as a downside but this needs to be further justified.

I agree with you that GPG is cumbersome, just think that this should be fully stated rather than assumed.

Reply:
i do think GPG should be allowed as an alternative source code signing mechanism (maybe in a different RFC), despite being cumbersome, since some people don't want to have downstream users have to trust that github or crates.io or other intermediaries were never compromised.

Reply:
I personally think that any solution like that should also work with "light" keys without the baggage of GPG, i.e. vanilla RSA/ED25519 keys like the OpenBSD signify tool or SSH. I definitely think that relying on the way GPG handles keys would be cumbersome, but having a decentralised method seems nice.